server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -464,6 +465,13 @@ private static SearchHits getHits(
                     assert shardDoc instanceof RankDoc;
                     searchHit.setRank(((RankDoc) shardDoc).rank);
                     searchHit.score(shardDoc.score);
+                    long shardAndDoc = ((long) shardDoc.shardIndex << 32) | (shardDoc.doc & 0xFFFFFFFFL);
```

Comment:
we should have a static universal method that does this encoding.

Reply:
Added in https://github.com/elastic/elasticsearch/pull/114085/commits/f90d278ce911119033e6957688381a8d342cbe0b